src/test/mir-opt/issue-91633.rs

```diff
@@ -0,0 +1,28 @@
+// EMIT_MIR issue_91633.hey.mir_map.0.mir
+fn hey<T> (it: &[T])
+ where
+     [T] : std::ops::Index<usize>,
+ {
+     let _ = &it[0];
+ }
+
+// EMIT_MIR issue_91633.bar.mir_map.0.mir
+fn bar<T> (it: Box<[T]>)
+ where
+     [T] : std::ops::Index<usize>,
+ {
+     let _ = it[0];
+ }
+
+// EMIT_MIR issue_91633.fun.mir_map.0.mir
+fn fun<T> (it: &[T])
+ {
+     let _ = &it[0];
+ }
+
+// EMIT_MIR issue_91633.foo.mir_map.0.mir
+fn foo<T> (it: Box<[T]>)
+ {
+     let _ = it[0];
+ }
```

Comment:
nit: this needs side effects for mir-opt not to remove the index call. Can you make all these fns return `-> &T` instead?